src/components/implementation/mem_mgr/parsec.c

```diff
@@ -0,0 +1,819 @@
+		last->next = freelist->head;
+		freelist->head = head;
+		freelist->n_items += return_n;
+		ck_spinlock_unlock(&(freelist->l));
+	}
+
+	return 0;
+}
+
+/* Add *new* items to the global freelist. no quiescence checked. */
+int 
+glb_freelist_add(void *item, struct parsec_allocator *alloc)
+{
+	struct freelist *freelist;
+	struct quie_mem_meta *meta;
+	
+	meta = item - sizeof(struct quie_mem_meta);
+
+	/* Should have been marked already. Set it anyway. */
+	meta->flags |= PARSEC_FLAG_DEACT;
+
+	freelist = &(alloc->glb_freelist.slab_freelists[size2slab(meta->size)]);
+
+	ck_spinlock_lock(&(freelist->l));
+	meta->next = freelist->head;
+	freelist->head = meta;
+	freelist->n_items++;
+	ck_spinlock_unlock(&(freelist->l));
+
+	return 0;
+}
+
+int
+parsec_free(void *node, struct parsec_allocator *alloc)
+{
+	struct quie_mem_meta *meta;
+	struct quie_queue *queue;
+	int cpu, old, new;
+
+	if (unlikely(!node || !alloc)) return -EINVAL;
+	/* if ((unsigned long)node % CACHE_LINE) return -EINVAL; */
+
+	cpu = get_cpu();
+	meta = (struct quie_mem_meta *)((unsigned long)node - sizeof(struct quie_mem_meta));
+	meta->time_deact = get_time();
+
+	old = meta->flags;
+	if (old & PARSEC_FLAG_DEACT) return -EINVAL;
+	new = meta->flags | PARSEC_FLAG_DEACT;
+
+	if (cos_cas((unsigned long *)(&meta->flags), old, new) != CAS_SUCCESS) return -ECASFAIL;
+
+	queue = &(alloc->qwq[cpu].slab_queue[size2slab(meta->size)]);
+	quie_queue_add(queue, meta);
+	
+	if (queue->n_items >= alloc->qwq_max_limit) {
+		quie_queue_balance(queue, alloc);
+	}
+	
+	return 0;
+}
+
+void *
+lib_exec(void *(*func)(void *), void *arg, parsec_t *p) 
+{
+	void *ret;
+
+	lib_enter(p);
+	ret = func(arg);
+	lib_exit(p);
+
+	return ret;
+}
+
+/* not used for now */
+static void parsec_quiesce(parsec_t *p)
+{
+	quie_time_t t = get_time();
+	
+	lib_enter(p);
+	parsec_quiescence_wait(t, p);
+	lib_exit(p);
+}
+
+/* static int  */
+/* qmem_free(void *mem) */
+/* { */
+/* 	int ret; */
+
+/* 	lib_enter(); */
+/* 	ret = q_free(mem); */
+/* 	lib_exit(); */
+
+/* 	return ret; */
+/* } */
+
+/* Not used for now. */
+/* This checks quiescence based on pure local knowledge (and without
+ * waiting for it). No remote readings. return 0 if quiesced. */
+/* static int */
+/* parsec_quiescence_quick_check(quie_time_t time_check) */
+/* { */
+/* 	struct quiescence_timing *timing_local; */
+
+/* 	timing_local = &timing_info[get_cpu()].timing; */
+/* 	if (time_check > timing_local->last_known_quiescence) return -1; */
+
+/* 	/\* quiesced *\/ */
+/* 	return 0; */
+/* } */
+
+void *
+timer_update(void *arg)
+{
+	(void)arg;
+	glb_ts.ts = 10;
+
+	/* printf("timer thread running on cpu %d\n", (int)arg); */
+	while (1) {
+		spin_delay(TS_GRANULARITY);
+		global_timestamp_inc();
+	}
+}
+
+struct thd_active {
+	int accessed;
+	int done;
+	int avg;
+	int max;
+	int read_avg;
+	int read_max;
+} CACHE_ALIGNED;
+
+struct thd_active thd_active[NUM_CPU] CACHE_ALIGNED;
+volatile int use_ncores;
+
+#ifdef IN_LINUX
```

Comment:
it is likely cleaner to move the linux-specific (and composite-specific) stuff into a separate file, and either compile the correct one in, or just use the `#ifdef` around the function call.